src/wasm/wasm-type.cpp

```diff
@@ -2807,16 +2868,18 @@ void ShapeCanonicalizer::translatePartitionsToTypes() {
           // Child doesn't need replacement.
           return;
         }
-        auto it = canonicalizer.states.find(*child);
-        if (it != canonicalizer.states.end()) {
-          // Child hasn't already been replaced; replace it.
-          auto set = canonicalizer.partitions.getSetForElem(it->second);
-          *child = canonicalizer.results.at(set.index);
-        }
+        canonicalizer.replaceHeapType(child);
       }
     };
     HeapType root = asHeapType(info);
     ChildUpdater(*this).walkRoot(&root);
+
+    // If this is a nominal type, we may need to update its supertype as well.
+    if (info->supertype) {
+      HeapType heapType(uintptr_t(info->supertype));
+      replaceHeapType(&heapType);
+      info->supertype = getHeapTypeInfo(heapType);
+    }
```

Comment:
Why is there no code for typesystem==nominal around here? (i.e., why was this not an issue before)

Reply:
`ShapeCanonicalizer` is also only used for equirecursive canonicalization, so `--nominal` mode never uses this code.